test/mpi/test_mpi_cuda.c

```diff
@@ -0,0 +1,531 @@
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    CUdevice cu_dev_other;
+
+    retain_and_push_primary_context(cu_dev);
+
+    alloc_mem_send = allocator_send->alloc(size);
+    alloc_mem_recv = allocator_recv->alloc(size);
+
+    CUDA_CALL(cuDeviceGetCount(&dev_count));
+
+    /* Send on contexts of all devices */
+    for (dev_idx = 1; dev_idx < dev_count; ++dev_idx) {
+        CUDA_CALL(cuDeviceGet(&cu_dev_other, dev_idx));
+        retain_and_push_primary_context(cu_dev_other);
+
+        res += mpi_pingpong(alloc_mem_send.ptr, alloc_mem_recv.ptr, size);
+
+        pop_and_release_primary_context(cu_dev_other);
+    }
+
+    allocator_send->free(&alloc_mem_send);
+    allocator_recv->free(&alloc_mem_recv);
+
+    pop_and_release_primary_context(cu_dev);
+
+    return res;
+}
+
+const test_t tests[] = {
+    {"MPI pingpong, memory allocation and communication are performed with the "
+     "same primary context set",
+     1, test_alloc_prim_send_prim},
+    {"MPI pingpong, memory is allocated with the primary context set, "
+     "communication is done with no context set",
+     1, test_alloc_prim_send_no},
+    {"MPI pingpong, memory is allocated with the primary context set, "
+     "communication is done with the user context set",
+     1, test_alloc_prim_send_user},
+    {"MPI pingpong, memory is allocated with the user context set, "
+     "communication is done with the primary context set",
+     1, test_alloc_user_send_prim},
+    {"MPI pingpong, memory is allocated with the primary context of GPU 0, "
+     "communication is done with the primary contexts of all other GPUs",
+     2, test_alloc_prim_send_others_prim}
+};
+
+static int check_allocator(const allocator_t *allocator, CUdevice cu_dev)
+{
+    alloc_mem_t alloc_mem;
+
+    retain_and_push_primary_context(cu_dev);
+    alloc_mem = allocator->alloc(1);
+    if (alloc_mem.ptr == 0) {
+        return 0;
+    }
+
+    allocator->free(&alloc_mem);
+    pop_and_release_primary_context(cu_dev);
+
+    return 1;
+}
+
+int main(int argc, char **argv)
+{
+    int comm_size;
+    int rank;
+    int dev_count, dev_idx;
+    CUdevice cu_dev;
+    CUcontext cu_ctx;
+    int i, j, k, l;
+    const test_t *test;
+    const allocator_t *allocator_send, *allocator_recv;
+    int res;
+
+    MPI_Init(&argc, &argv);
+    MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
+    if (comm_size != 2) {
+        PRINT_ROOT("This test requires 2 processes, not %d\n", comm_size);
+        goto out;
+    }
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    MPI_Comm_set_errhandler(MPI_COMM_WORLD, MPI_ERRORS_RETURN);
+
+    CUDA_CALL(cuInit(0));
+    CUDA_CALL(cuDeviceGetCount(&dev_count));
+    if (dev_count < 1) {
+        PRINT_ROOT("This test requires at least 1 GPU\n");
+        goto out;
+    }
+
+    for (dev_idx = dev_count - 1; dev_idx > -1; --dev_idx) {
+        CUDA_CALL(cuDeviceGet(&cu_dev, dev_idx));
+        CUDA_CALL(cuDevicePrimaryCtxRetain(&cu_ctx, cu_dev));
+    }
```

Comment:
@rakhmets Minor comment -- By this logic, if the MPI job were to be launched on the same node, both ranks would end up initializing cu_dev to 0 by the end of the above loop. If there are multiple GPUs on the node, I think it would be good to check where the ranks manage different GPUs. As an example, osu benchmarks use LOCAL_RANK env for this reason.